# iNaturalist status updates by state - QLD

Using the file produced in the collate-status-taxa.ipynb: `inat-aust-status-taxa.csv`, generate lists to update iNaturalist statuses

**Next steps:**
Establish the changes that need to be made. Read in the sensitive and conservation list for each state.
    a. new - any new species that appear in the state lists but do not have a status in inaturalist (new template)
    b. updates - any changes to information which was added by us previously (user_id = 708886) (update template, action='UPDATE')
    c. removals - any statuses which were added by us previously (user_id = 708886) list which are incorrect (update template, action='REMOVE')
    d. flags - are there any statuses by other users that need to be flagged?

## Prep - common to all states
1. Read in the inaturalist statuses & filter out this state
2. Read in the inaturalist taxa list
3. Read in the state sensitive and conservation list
4. Attempt to match the state statuses to an IUCN equivalent
5. Determine the best placeID to use for this state


### 1. iNaturalist statuses

In [28]:
import pandas as pd
import sys
import os
projectdir = "/Users/oco115/PycharmProjects/authoritative-lists/" # basedir for this gh project
sys.path.append(os.path.abspath(projectdir + "source-code/includes"))
import list_functions as lf

# projectdir = "/Users/new330/IdeaProjects/authoritative-lists/" # basedir for this gh project
sourcedir = projectdir + "source-data/inaturalist-statuses/"
listdir = projectdir + "current-lists/"


# read in the statuses
taxastatus = pd.read_csv(sourcedir + "inat-aust-status-taxa.csv", encoding='UTF-8',na_filter=False,dtype=str) ## Read inaturalist conservation statuses file
taxastatus.head(3)

id taxon_id  user_id place_id source_id  \
0  166449    38493  1138587     7830             
1  234788   918383   702203     9994             
2  234789   918383   702203     7308             

                            authority status  \
0  Flora and Fauna Guarantee Act 1988     CR   
1           Atlas of Living Australia    NT    
2           Atlas of Living Australia     LC   

                                                 url description geoprivacy  \
0                                                                  obscured   
1  https://bie.ala.org.au/species/https://id.biod...                          
2  https://bie.ala.org.au/species/https://id.biod...                          

   ...         genus specificEpithet infraspecificEpithet  \
0  ...     Eulamprus      kosciuskoi                        
1  ...  Chiloschista     phyllorhiza                        
2  ...  Chiloschista     phyllorhiza                        

               modified            scientificName taxonRank  \
0  2021-03-01T10:35:01Z      Eulamprus kosciuskoi   species   
1  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   
2  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   

                                          references preferred_common_name  \
0  http://reptile-database.reptarium.cz/search.ph...                         
1  http://www.catalogueoflife.org/annual-checklis...                         
2  http://www.catalogueoflife.org/annual-checklis...                         

  is_active current_synonymous_taxon_ids  
0                                         
1                                         
2                                         

[3 rows x 34 columns]

In [29]:
def filter_state_statuses(stateregex: str, urlregex: str):
    authoritydf = taxastatus['authority'].drop_duplicates().sort_values()
    authoritydf = authoritydf[pd.Series(authoritydf).str.contains(stateregex)]
    urldf = taxastatus['url'].drop_duplicates().sort_values()
    urldf = urldf[pd.Series(urldf).str.contains(urlregex)]
    placedisplaydf = taxastatus['place_display_name'].drop_duplicates().sort_values()
    placedisplaydf = placedisplaydf[pd.Series(placedisplaydf).str.contains(stateregex)]
    placedf = taxastatus['place_name'].drop_duplicates().sort_values()
    placedf = placedf[pd.Series(placedf).str.contains(stateregex)]
    # concat all and remove duplicates
    statedf = pd.concat([taxastatus.apply(lambda row: row[taxastatus['place_display_name'].isin(placedisplaydf)]),
                         taxastatus.apply(lambda row: row[taxastatus['place_name'].isin(placedf)]),
                         taxastatus.apply(lambda row: row[taxastatus['url'].isin(urldf)]),
                         taxastatus.apply(
                             lambda row: row[taxastatus['authority'].isin(authoritydf)])]).drop_duplicates()
    return statedf.sort_values(['taxon_id', 'user_id'])


inatstatuses = filter_state_statuses("Qld|QLD|Queensland|QUEENSLAND|QL", ".qld.")
inatstatuses.rename(columns={'id':'status_id','id_y':'taxon_id_y'},inplace=True)
inatstatuses

status_id taxon_id  user_id place_id source_id  \
532     223098  1003032  3669610     7308             
131     169944  1004434    58320     7308             
1473    152757   100615   708886     7308     16653   
1407    152668   100635   708886     7308     16653   
1267    152461   100638   708886     7308     16653   
...        ...      ...      ...      ...       ...   
835     262047   977629  3669610     7308             
2615    161807   993333   702203     7308             
2616    161808   993333   702203     6744             
23      167723   993605  3669610     7308             
2897    164338   993923    58320     7308             

                                              authority           status  \
532             Queensland Nature Conservation Act 1992               CE   
131             Queensland Nature Conservation Act 1992    Least concern   
1473                                           QLD DEHP       endangered   
1407                                           QLD DEHP       endangered   
1267                                           QLD DEHP       endangered   
...                                                 ...              ...   
835             Queensland Nature Conservation Act 1992               NT   
2615                      Nature Conservation Act 1992                VU   
2616  Environment Protection and Biodiversity Conser...               VU   
23                                             QLD DEHP               NT   
2897            Queensland Nature Conservation Act 1992  Near threatened   

                                                    url  \
532   https://apps.des.qld.gov.au/species-search/det...   
131   https://apps.des.qld.gov.au/species-search/det...   
1473  https://data.qld.gov.au/dataset/conservation-s...   
1407  https://data.qld.gov.au/dataset/conservation-s...   
1267  https://data.qld.gov.au/dataset/conservation-s...   
...                                                 ...   
835   https://www.data.qld.gov.au/dataset/conservati...   
2615  https://apps.des.qld.gov.au/species-search/det...   
2616  https://apps.des.qld.gov.au/species-search/det...   
23    https://www.data.qld.gov.au/dataset/conservati...   
2897  https://apps.des.qld.gov.au/species-search/det...   

                                            description geoprivacy  ...  \
532                                                       obscured  ...   
131   Listed as 'confidential' by the Queensland Dep...   obscured  ...   
1473                                                      obscured  ...   
1407                                                      obscured  ...   
1267                                                      obscured  ...   
...                                                 ...        ...  ...   
835                                                           open  ...   
2615                                                                ...   
2616                                                                ...   
23                                                        obscured  ...   
2897       See https://www.inaturalist.org/flags/512314       open  ...   

            genus       specificEpithet infraspecificEpithet  \
532    Antrophyum  austroqueenslandicum                        
131      Oberonia         crateriformis                        
1473    Euastacus                bindal                        
1407    Euastacus                jagara                        
1267    Euastacus                maidae                        
...           ...                   ...                  ...   
835     Melaleuca               formosa                        
2615  Cupaniopsis            tomentella                        
2616  Cupaniopsis            tomentella                        
23      Acianthus         amplexicaulis                        
2897      Ardisia                bakeri                        

                  modified                   scienti

### 2. iNaturalist taxonomy

In [30]:
# Output files contain these fields
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# so we need to match species from the state lists to the inat taxa to get the taxon_id

import zipfile
url = "https://www.inaturalist.org/taxa/inaturalist-taxonomy.dwca.zip"
filename = url.split("/")[-1]

z=zipfile.ZipFile(sourcedir + filename)

with z.open('taxa.csv') as from_archive:
    inattaxa = pd.read_csv(from_archive,dtype=str)
z.close()
inattaxa.head(3)


id                             taxonID                          identifier  \
0  1  https://www.inaturalist.org/taxa/1  https://www.inaturalist.org/taxa/1   
1  2  https://www.inaturalist.org/taxa/2  https://www.inaturalist.org/taxa/2   
2  3  https://www.inaturalist.org/taxa/3  https://www.inaturalist.org/taxa/3   

                         parentNameUsageID   kingdom    phylum class order  \
0   https://www.inaturalist.org/taxa/48460  Animalia       NaN   NaN   NaN   
1       https://www.inaturalist.org/taxa/1  Animalia  Chordata   NaN   NaN   
2  https://www.inaturalist.org/taxa/355675  Animalia  Chordata  Aves   NaN   

  family genus specificEpithet infraspecificEpithet              modified  \
0    NaN   NaN             NaN                  NaN  2021-11-02T06:05:44Z   
1    NaN   NaN             NaN                  NaN  2021-11-23T00:40:18Z   
2    NaN   NaN             NaN                  NaN  2022-12-27T07:33:16Z   

  scientificName taxonRank                                         references  
0       Animalia   kingdom  http://www.catalogueoflife.org/annual-checklis...  
1       Chordata    phylum  http://www.catalogueoflife.org/annual-checklis...  
2           Aves     class  http://www.catalogueoflife.org/annual-checklis...

### 3. State lists

Sensitive list: `geoprivacy` = `obscured`
overrides


In [31]:
# sensitivelist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr493?max=10000&includeKVP=true")
sensitivelist = lf.download_ala_list("https://lists-test.ala.org.au/ws/speciesListItems/dr18404?max=10000&includeKVP=true")
sensitivelist = lf.kvp_to_columns(sensitivelist)
sensitivelist['name'] = sensitivelist['name'].str.replace('subsp. ', '', regex=False)
sensitivelist = sensitivelist.rename(columns={'taxonID':'wildnetTaxonID'})
sensitivelist['wildnet_geoprivacy'] = 'obscured'
sensitivelist

id                          name  \
0   2799651    Rhinolophus philippinensis   
0   2799225             Chloebia gouldiae   
0   2798738            Erythrura trichroa   
0   2799209  Neochmia phaeton evangelinae   
0   2799217        Poephila cincta cincta   
..      ...                           ...   
0   2799125        Pneumatopteris costata   
0   2799126      Pneumatopteris pennigera   
0   2799558         Thelypteris confluens   
0   2799401            Macadamia jansenii   
0   2799588        Tmesipteris lanceolata   

                                    commonName  \
0            Greater Large-eared Horseshoe Bat   
0                               Gouldian Finch   
0                      Blue-faced Parrot-finch   
0                                         None   
0   Black-throated Finch (southern Subspecies)   
..                                         ...   
0                                         None   
0                                        Pākau   
0                                   Swamp Fern   
0                            Bulburin Nut Tree   
0                                         None   

                             scientificName  \
0                Rhinolophus philippinensis   
0                         Chloebia gouldiae   
0            Erythrura (Acalanthe) trichroa   
0   Neochmia (Neochmia) phaeton evangelinae   
0         Poephila (Poephila) cincta cincta   
..                                      ...   
0                    Pneumatopteris costata   
0                  Pneumatopteris pennigera   
0                     Thelypteris confluens   
0                        Macadamia jansenii   
0                    Tmesipteris lanceolata   

                                                 lsid dataResourceUid  \
0                                       ALA_DR652_255         dr18404   
0   https://biodiversity.org.au/afd/taxa/27d27be2-...         dr18404   
0   https://biodiversity.org.au/afd/taxa/cb632014-...         dr18404   
0   https://biodiversity.org.au/afd/taxa/565b6f37-...         dr18404   
0   https://biodiversity.org.au/afd/taxa/9fe878db-...         dr18404   
..                                                ...             ...   
0    https://id.biodiversity.org.au/node/apni/2887362         dr18404   
0    https://id.biodiversity.org.au/node/apni/2899796         dr18404   
0    https://id.biodiversity.org.au/node/apni/2913519         dr18404   
0    https://id.biodiversity.org.au/node/apni/2904937         dr18404   
0    https://id.biodiversity.org.au/node/apni/2907902         dr18404   

                                            kvpValues wildnetTaxonID  \
0   [{'key': 'taxonID', 'value': '969'}, {'key': '...            969   
0   [{'key': 'taxonID', 'value': '1376'}, {'key': ...           1376   
0   [{'key': 'taxonID', 'value': '1378'}, {'key': ...           1378   
0   [{'key': 'taxonID', 'value': '1370'}, {'key': ...           1370   
0   [{'key': 'taxonID', 'value': '1365'}, {'key': ...           1365   
..                                                ...            ...   
0   [{'key': 'taxonID', 'value': '11699'}, {'key':...          11699   
0   [{'key': 'taxonID', 'value': '11700'}, {'key':...          11700   
0   [{'key': 'taxonID', 'value': '16042'}, {'key':...          16042   
0   [{'key': 'taxonID', 'value': '8185'}, {'key': ...           8185   
0   [{'key': 'taxonID', 'value': '8486'}, {'key': ...           8486   

       scientificNameAuthorship sourceStatus Significant  \
0              Waterhouse, 1843            E           Y   
0                 (Gould, 1844)            E           Y   
0              (Kittlitz, 1835)           NT           Y   
0   (Hombron & Jacquinot, 1841)            E           Y   
0                   Gould, 1837            E           Y   
..                          ...          ...         ...   
0              (Brack.) Holttum           NT           Y   
0            (G.Forst.) Holttum            E           Y   
0           (Thunb.)

In [32]:
conservationlist = lf.download_ala_list("https://lists-test.ala.org.au/ws/speciesListItems/dr652?max=10000&includeKVP=true")
conservationlist = lf.kvp_to_columns(conservationlist)
conservationlist['name'] = conservationlist['name'].str.replace('subsp. ', '', regex=False)
conservationlist = conservationlist.rename(columns={'taxonID':'wildnetTaxonID'})
conservationlist['wildnet_geoprivacy'] = conservationlist['Confidential'].apply(lambda x: 'obscured' if x == 'Y' else 'open')
conservationlist

id                   name           commonName  \
0   2798372        Adelotus brevis          Tusked Frog   
0   2796770   Philoria kundagungan        Mountain Frog   
0   2796819         Crinia tinnula       Wallum Froglet   
0   2797552       Mixophyes fleayi  Fleay's Barred Frog   
0   2796637     Mixophyes iteratus    Giant Barred Frog   
..      ...                    ...                  ...   
0   2797742     Macrozamia viridis                 None   
0   2796733       Alpinia hylandii                 None   
0   2797715  Amomum queenslandicum                 None   
0   2797487       Globba marantina            Kera Kera   
0   2796825     Zostera capricorni                 None   

                        scientificName  \
0                      Adelotus brevis   
0                 Philoria kundagungan   
0                       Crinia tinnula   
0                     Mixophyes fleayi   
0                   Mixophyes iteratus   
..                                 ...   
0                   Macrozamia viridis   
0                     Alpinia hylandii   
0                Amomum queenslandicum   
0                     Globba marantina   
0   Zostera muelleri subsp. capricorni   

                                                 lsid dataResourceUid  \
0   https://biodiversity.org.au/afd/taxa/32f57cc7-...           dr652   
0   https://biodiversity.org.au/afd/taxa/fcb2e700-...           dr652   
0   https://biodiversity.org.au/afd/taxa/a101af4a-...           dr652   
0   https://biodiversity.org.au/afd/taxa/289e1b7c-...           dr652   
0   https://biodiversity.org.au/afd/taxa/9eef5215-...           dr652   
..                                                ...             ...   
0    https://id.biodiversity.org.au/node/apni/2917053           dr652   
0    https://id.biodiversity.org.au/node/apni/2907514           dr652   
0    https://id.biodiversity.org.au/node/apni/2893545           dr652   
0    https://id.biodiversity.org.au/node/apni/2901385           dr652   
0    https://id.biodiversity.org.au/node/apni/2905201           dr652   

                                            kvpValues wildnetTaxonID  \
0   [{'key': 'taxonID', 'value': '706'}, {'key': '...            706   
0   [{'key': 'taxonID', 'value': '687'}, {'key': '...            687   
0   [{'key': 'taxonID', 'value': '686'}, {'key': '...            686   
0   [{'key': 'taxonID', 'value': '675'}, {'key': '...            675   
0   [{'key': 'taxonID', 'value': '676'}, {'key': '...            676   
..                                                ...            ...   
0   [{'key': 'taxonID', 'value': '6482'}, {'key': ...           6482   
0   [{'key': 'taxonID', 'value': '8948'}, {'key': ...           8948   
0   [{'key': 'taxonID', 'value': '8949'}, {'key': ...           8949   
0   [{'key': 'taxonID', 'value': '12434'}, {'key':...          12434   
0   [{'key': 'taxonID', 'value': '14130'}, {'key':...          14130   

   scientificNameAuthorship sourceStatus Significant Confidential  \
0           (Günther, 1863)            V           Y            N   
0   (Ingram & Corben, 1975)            E           Y            Y   
0    Straughan & Main, 1966            V           Y            N   
0     Corben & Ingram, 1987            E           Y            Y   
0           Straughan, 1968            V           Y            Y   
..                      ...          ...         ...          ...   
0    D.L.Jones & P.I.Forst.            E           Y            Y   
0                   R.M.Sm.           NT           Y            N   
0                   R.M.Sm.            V           Y            N   
0                        L.            V           Y            N   
0                     Asch.           SL           Y            N   

                   status          Endemicity E P B C  Status  \
0              Vulnerable       Intranational             NaN   
0              Endangered       Intranational      Endangered   
0              Vulnerable    

In [33]:
# join them in a way that works for inat (eg sensitive list, geoprivacy = 'obscured'
statelist = pd.concat([sensitivelist[['wildnetTaxonID', 'name', 'status', 'wildnet_geoprivacy', 'lsid']],
                    conservationlist[['wildnetTaxonID', 'name', 'status', 'wildnet_geoprivacy', 'lsid']]]).drop_duplicates()
# retrieve binomial and trinomial names from GBIF
parsednames = lf.gbifparse(statelist)
parsednames.to_csv(sourcedir + "qld-gbif.csv", index=False)
statelist = statelist.merge(parsednames[['scientificName','canonicalName','canonicalNameComplete','type','rankMarker']],how="left",left_on="name",right_on="scientificName")
numfullstatelist = len(statelist.index)
statelist['scientificName'] = statelist['canonicalName']
statelist

wildnetTaxonID                                               name  \
0               969                         Rhinolophus philippinensis   
1              1376                                  Chloebia gouldiae   
2              1378                                 Erythrura trichroa   
3              1370                       Neochmia phaeton evangelinae   
4              1365                             Poephila cincta cincta   
...             ...                                                ...   
2512           6296  Xanthorrhoea sp. (Cape Bedford M.Gandini AQ601...   
2513           8948                                   Alpinia hylandii   
2514           8949                              Amomum queenslandicum   
2515          12434                                   Globba marantina   
2516          14130                                 Zostera capricorni   

                     status wildnet_geoprivacy  \
0                Endangered           obscured   
1                Endangered           obscured   
2           Near Threatened           obscured   
3                Endangered           obscured   
4                Endangered           obscured   
...                     ...                ...   
2512  Special least concern               open   
2513        Near Threatened               open   
2514             Vulnerable               open   
2515             Vulnerable               open   
2516  Special least concern               open   

                                                   lsid  \
0                                         ALA_DR652_255   
1     https://biodiversity.org.au/afd/taxa/27d27be2-...   
2     https://biodiversity.org.au/afd/taxa/cb632014-...   
3     https://biodiversity.org.au/afd/taxa/565b6f37-...   
4     https://biodiversity.org.au/afd/taxa/9fe878db-...   
...                                                 ...   
2512   https://id.biodiversity.org.au/node/apni/2906735   
2513   https://id.biodiversity.org.au/node/apni/2907514   
2514   https://id.biodiversity.org.au/node/apni/2893545   
2515   https://id.biodiversity.org.au/node/apni/2901385   
2516   https://id.biodiversity.org.au/node/apni/2905201   

                    scientificName                 canonicalName  \
0       Rhinolophus philippinensis    Rhinolophus philippinensis   
1                Chloebia gouldiae             Chloebia gouldiae   
2               Erythrura trichroa            Erythrura trichroa   
3     Neochmia phaeton evangelinae  Neochmia phaeton evangelinae   
4           Poephila cincta cincta        Poephila cincta cincta   
...                            ...                           ...   
2512            Xanthorrhoea spec.            Xanthorrhoea spec.   
2513              Alpinia hylandii              Alpinia hylandii   
2514         Amomum queenslandicum         Amomum queenslandicum   
2515              Globba marantina              Globba marantina   
2516            Zostera capricorni            Zostera capricorni   

             canonicalNameComplete        type rankMarker  
0       Rhinolophus philippinensis  SCIENTIFIC        sp.  
1                Chloebia gouldiae  SCIENTIFIC        sp.  
2               Erythrura trichroa  SCIENTIFIC        sp.  
3     Neochmia phaeton evangelinae  SCIENTIFIC   infrasp.  
4           Poephila cincta cincta  SCIENTIFIC   infrasp.  
...                            ...         ...        ...  
2512            Xanthorrhoea spec.    INFORMAL        sp.  
2513              Alpinia hylandii  SCIENTIFIC        sp.  
2514         Amomum queenslandicum  SCIENTIFIC        sp.  
2515              Globba marantina  SCIENTIFIC        sp.  
2516            Zostera capricorni  SCIENTIFIC        sp.  

[2517 rows x 10 columns]

In [34]:
# Identify records that won't comply with iNaturalist species names
noncomply = statelist[statelist['type'].isin(['INFORMAL','CULTIVAR','HYBRID']) ]

In [35]:
# remove records that do not comply
statelist = statelist[~statelist['type'].isin(['INFORMAL','CULTIVAR','HYBRID']) ]
statelist = pd.DataFrame(statelist[['wildnetTaxonID','scientificName','status','wildnet_geoprivacy','lsid']])
statelist

wildnetTaxonID                scientificName                 status  \
0               969    Rhinolophus philippinensis             Endangered   
1              1376             Chloebia gouldiae             Endangered   
2              1378            Erythrura trichroa        Near Threatened   
3              1370  Neochmia phaeton evangelinae             Endangered   
4              1365        Poephila cincta cincta             Endangered   
...             ...                           ...                    ...   
2511          12306          Xanthorrhoea pumilio  Special least concern   
2513           8948              Alpinia hylandii        Near Threatened   
2514           8949         Amomum queenslandicum             Vulnerable   
2515          12434              Globba marantina             Vulnerable   
2516          14130            Zostera capricorni  Special least concern   

     wildnet_geoprivacy                                               lsid  
0              obscured                                      ALA_DR652_255  
1              obscured  https://biodiversity.org.au/afd/taxa/27d27be2-...  
2              obscured  https://biodiversity.org.au/afd/taxa/cb632014-...  
3              obscured  https://biodiversity.org.au/afd/taxa/565b6f37-...  
4              obscured  https://biodiversity.org.au/afd/taxa/9fe878db-...  
...                 ...                                                ...  
2511               open   https://id.biodiversity.org.au/node/apni/2916574  
2513               open   https://id.biodiversity.org.au/node/apni/2907514  
2514               open   https://id.biodiversity.org.au/node/apni/2893545  
2515               open   https://id.biodiversity.org.au/node/apni/2901385  
2516               open   https://id.biodiversity.org.au/node/apni/2905201  

[2473 rows x 5 columns]

In [36]:
parsednames['type'].unique()

array(['SCIENTIFIC', 'HYBRID', 'INFORMAL', 'CULTIVAR', 'PLACEHOLDER'],
      dtype=object)

In [37]:
# check for duplicates with conflicting information
dupinformation = statelist.groupby('wildnetTaxonID').filter(lambda x: len(x) > 1)#.sort('size',ascending=False)
dupinformation

Empty DataFrame
Columns: [wildnetTaxonID, scientificName, status, wildnet_geoprivacy, lsid]
Index: []

### 4. Equivalent IUCN statuses

In [38]:
# iucn_statuses = {'Not Evaluated', 'Data Deficient', 'Least Concern', 'Near Threatened', 'Vulnerable', 'Endangered', 'Critically Endangered', 'Extinct in the Wild','Extinct'}
statelist.groupby(['status'])['status'].count()

status
Critically Endangered    132
Endangered               288
Extinct                    5
Extinct in the wild       26
Least concern            132
Near Threatened          273
Special least concern    900
Vulnerable               593
Name: status, dtype: int64

In [39]:
# iucnStatusMappings = {

iucnStatusMappings = {
    'Least concern':'Least Concern',
    'Special least concern':'Least Concern',
    'Extinct in the wild':'Extinct in the Wild'
}

### 5. Determine best place ID to use

In [40]:
inatstatuses.groupby(['place_id','place_name','place_display_name'])['place_id'].count()
# looks like 7308

place_id  place_name             place_display_name           
                                                                    5
144315    Brisbane City          Brisbane City                      2
153119    South East Queensland  South East Queensland, QL, AU      1
18870     Cairns - Pt B          Cairns - Pt B, QL, AU              1
19232     Yarrabah               Yarrabah, QL, AU                   1
6744      Australia              Australia                         12
7308      Queensland             Queensland, AU                   631
Name: place_id, dtype: int64

## Merge iNaturalist statuses with State lists on scientificName

1. Match - updates, even if the statuses are the same we'll update the links and values anyway
2. No match - statuses to be added (additions)
   1.1 No match and no taxnomy - search for synonyms
   1.2 No match
3. Merge the other direction to see if there are deletes?


In [41]:
# join to see which lists already have a status in inaturalist based on scientificName
mergedstatuses = statelist[['wildnetTaxonID','scientificName','status','wildnet_geoprivacy', 'lsid']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='scientificName',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['scientificName'])
mergedstatuses


wildnetTaxonID        scientificName                 status  \
1651          33360   Abrodictyum brassii  Special least concern   
1652          33358  Abrodictyum caudatum  Special least concern   
1653          33359  Abrodictyum obscurum  Special least concern   
1743          11892    Acacia acrionastes        Near Threatened   
1744          14898      Acacia ammophila             Vulnerable   
...             ...                   ...                    ...   
2299          28217         Zieria vagans  Critically Endangered   
2300           3296      Zieria verrucosa             Vulnerable   
2301          41054    Zieria wilhelminae  Critically Endangered   
2483          14130    Zostera capricorni  Special least concern   
1127            876                   NaN             Endangered   

     wildnet_geoprivacy                                               lsid  \
1651               open   https://id.biodiversity.org.au/node/apni/7401847   
1652               open   https://id.biodiversity.org.au/node/apni/7402200   
1653               open   https://id.biodiversity.org.au/node/apni/7402565   
1743               open   https://id.biodiversity.org.au/node/apni/2900521   
1744               open   https://id.biodiversity.org.au/node/apni/2899480   
...                 ...                                                ...   
2299               open   https://id.biodiversity.org.au/node/apni/2900138   
2300               open  https://id.biodiversity.org.au/taxon/apni/5136...   
2301               open                                     ALA_DR652_2004   
2483               open   https://id.biodiversity.org.au/node/apni/2905201   
1127               open                                      ALA_DR652_233   

     status_id taxon_id user_id description iucn authority status_inat  \
1651       NaN      NaN     NaN         NaN  NaN       NaN         NaN   
1652       NaN      NaN     NaN         NaN  NaN       NaN         NaN   
1653       NaN      NaN     NaN         NaN  NaN       NaN         NaN   
1743       NaN      NaN     NaN         NaN  NaN       NaN         NaN   
1744       NaN      NaN     NaN         NaN  NaN       NaN         NaN   
...        ...      ...     ...         ...  ...       ...         ...   
2299       NaN      NaN     NaN         NaN  NaN       NaN         NaN   
2300       NaN      NaN     NaN         NaN  NaN       NaN         NaN   
2301       NaN      NaN     NaN         NaN  NaN       NaN         NaN   
2483       NaN      NaN     NaN         NaN  NaN       NaN         NaN   
1127       NaN      NaN     NaN         NaN  NaN       NaN         NaN   

     geoprivacy place_id place_display_name  
1651        NaN      NaN                NaN  
1652        NaN      NaN                NaN  
1653        NaN      NaN                NaN  
1743        NaN      NaN                NaN  
1744        NaN      NaN                NaN  
...         ...      ...                ...  
2299        NaN      NaN                NaN  
2300        NaN      NaN                NaN  
2301        NaN      NaN                NaN  
2483        NaN      NaN                NaN  
1127        NaN      NaN                NaN  

[2484 rows x 15 columns]

In [42]:
# prepare the export fields, common to New template and Update template
# new statuses
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username
mergedstatuses['new_authority'] = "Qld Department of Environment and Science"
mergedstatuses['new_description'] = "Listed as Confidential - refer to https://www.data.qld.gov.au/dataset/queensland-confidential-species"
mergedstatuses['new_url'] = mergedstatuses['lsid']
# mergedstatuses['new_url'] = "https://apps.des.qld.gov.au/species-search/details/?id=" + mergedstatuses['wildnetTaxonID'].astype(str)
mergedstatuses.rename(columns={'wildnet_geoprivacy':'new_geoprivacy'},inplace=True)
mergedstatuses['new_place_id'] = '7308'  # Queensland, AU
mergedstatuses['new_username'] = 'peggydnew'
mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
mergedstatuses['new_status'] = mergedstatuses['status'].fillna('Confidential')
mergedstatuses

wildnetTaxonID        scientificName                 status  \
1651          33360   Abrodictyum brassii  Special least concern   
1652          33358  Abrodictyum caudatum  Special least concern   
1653          33359  Abrodictyum obscurum  Special least concern   
1743          11892    Acacia acrionastes        Near Threatened   
1744          14898      Acacia ammophila             Vulnerable   
...             ...                   ...                    ...   
2299          28217         Zieria vagans  Critically Endangered   
2300           3296      Zieria verrucosa             Vulnerable   
2301          41054    Zieria wilhelminae  Critically Endangered   
2483          14130    Zostera capricorni  Special least concern   
1127            876                   NaN             Endangered   

     new_geoprivacy                                               lsid  \
1651           open   https://id.biodiversity.org.au/node/apni/7401847   
1652           open   https://id.biodiversity.org.au/node/apni/7402200   
1653           open   https://id.biodiversity.org.au/node/apni/7402565   
1743           open   https://id.biodiversity.org.au/node/apni/2900521   
1744           open   https://id.biodiversity.org.au/node/apni/2899480   
...             ...                                                ...   
2299           open   https://id.biodiversity.org.au/node/apni/2900138   
2300           open  https://id.biodiversity.org.au/taxon/apni/5136...   
2301           open                                     ALA_DR652_2004   
2483           open   https://id.biodiversity.org.au/node/apni/2905201   
1127           open                                      ALA_DR652_233   

     status_id taxon_id user_id description iucn  ... geoprivacy place_id  \
1651       NaN      NaN     NaN         NaN  NaN  ...        NaN      NaN   
1652       NaN      NaN     NaN         NaN  NaN  ...        NaN      NaN   
1653       NaN      NaN     NaN         NaN  NaN  ...        NaN      NaN   
1743       NaN      NaN     NaN         NaN  NaN  ...        NaN      NaN   
1744       NaN      NaN     NaN         NaN  NaN  ...        NaN      NaN   
...        ...      ...     ...         ...  ...  ...        ...      ...   
2299       NaN      NaN     NaN         NaN  NaN  ...        NaN      NaN   
2300       NaN      NaN     NaN         NaN  NaN  ...        NaN      NaN   
2301       NaN      NaN     NaN         NaN  NaN  ...        NaN      NaN   
2483       NaN      NaN     NaN         NaN  NaN  ...        NaN      NaN   
1127       NaN      NaN     NaN         NaN  NaN  ...        NaN      NaN   

     place_display_name                              new_authority  \
1651                NaN  Qld Department of Environment and Science   
1652                NaN  Qld Department of Environment and Science   
1653                NaN  Qld Department of Environment and Science   
1743                NaN  Qld Department of Environment and Science   
1744                NaN  Qld Department of Environment and Science   
...                 ...                                        ...   
2299                NaN  Qld Department of Environment and Science   
2300                NaN  Qld Department of Environment and Science   
2301                NaN  Qld Department of Environment and Science   
2483                NaN  Qld Department of Environment and Science   
1127                NaN  Qld Department of Environment and Science   

                                        new_description  \
1651  Listed as Confidential - refer to https://www....   
1652  Listed as Confidential - refer to https://www....   
1653  Listed as Confidential - refer to https://www....   
1743  Listed as Confidential - refer to https://www....   
1744  Listed as Confidential - refer to https://www....   
...                                                 ...   
2299  Listed as Confidential - refer to https://www....   
2300  Listed as Confidential - refer to https://www....   
23

## Updates

In [43]:
# those that need to be updated - we found a status
mergedstatuses[mergedstatuses['status_id'].notnull()][['scientificName','status','status_inat','new_geoprivacy','geoprivacy','authority','user_id']]

scientificName                 status status_inat new_geoprivacy  \
1749      Acacia attenuata             Vulnerable          VU           open   
1750   Acacia barakulensis             Vulnerable          VU           open   
1751  Acacia baueri baueri             Vulnerable          VU           open   
1752       Acacia calantha        Near Threatened         NT            open   
1761     Acacia hockingsii             Vulnerable          VU           open   
...                    ...                    ...         ...            ...   
2283         Zieria bifida             Endangered          EN           open   
2286        Zieria collina             Vulnerable          VU           open   
2288          Zieria exsul  Critically Endangered          CR           open   
2291     Zieria gymnocarpa  Critically Endangered          CR           open   
2298       Zieria scopulus  Critically Endangered          CR           open   

     geoprivacy                                authority  user_id  
1749                                          Queensland   702203  
1750   obscured                                 QLD DEHP   702203  
1751                                          Queensland   702203  
1752                               Queensland Government   702203  
1761       open  Queensland Nature Conservation Act 1992  3669610  
...         ...                                      ...      ...  
2283             Queensland Nature Conservation Act 1992   702203  
2286       open             Nature Conservation Act 1992   702203  
2288                                          Queensland   702203  
2291                                            QLD DEHP   409010  
2298             Queensland Nature Conservation Act 1992    58320  

[574 rows x 7 columns]

In [44]:
# updates - create the data frame
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username,description
updates = pd.DataFrame(mergedstatuses[mergedstatuses['status_id'].notnull()])
updates.sort_values('scientificName')
updates['action'] = 'UPDATE'
#updates.loc[:,'action'] = 'UPDATE'
updates = updates[['action','scientificName','status_id','taxon_id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
updates.columns = updates.columns.str.replace("new_", "", regex=True)
updates = updates.rename(columns={'scientificName':'taxon_name',
                                  'status_id':'id'})
updates

action            taxon_name      id taxon_id                 status  \
1749  UPDATE      Acacia attenuata  159918   898644             Vulnerable   
1750  UPDATE   Acacia barakulensis  158182  1038687             Vulnerable   
1751  UPDATE  Acacia baueri baueri  159921  1111700             Vulnerable   
1752  UPDATE       Acacia calantha  160798  1121183        Near Threatened   
1761  UPDATE     Acacia hockingsii  264938  1252519             Vulnerable   
...      ...                   ...     ...      ...                    ...   
2283  UPDATE         Zieria bifida  223195  1125311             Endangered   
2286  UPDATE        Zieria collina  262040   537414             Vulnerable   
2288  UPDATE          Zieria exsul  159923  1111558  Critically Endangered   
2291  UPDATE     Zieria gymnocarpa  169061  1244660  Critically Endangered   
2298  UPDATE       Zieria scopulus  168010  1097883  Critically Endangered   

     iucn_equivalent                                  authority  \
1749      Vulnerable  Qld Department of Environment and Science   
1750      Vulnerable  Qld Department of Environment and Science   
1751      Vulnerable  Qld Department of Environment and Science   
1752      Vulnerable  Qld Department of Environment and Science   
1761      Vulnerable  Qld Department of Environment and Science   
...              ...                                        ...   
2283      Vulnerable  Qld Department of Environment and Science   
2286      Vulnerable  Qld Department of Environment and Science   
2288      Vulnerable  Qld Department of Environment and Science   
2291      Vulnerable  Qld Department of Environment and Science   
2298      Vulnerable  Qld Department of Environment and Science   

                                                   url geoprivacy place_id  \
1749  https://id.biodiversity.org.au/node/apni/2887463       open     7308   
1750  https://id.biodiversity.org.au/node/apni/2903278       open     7308   
1751  https://id.biodiversity.org.au/node/apni/2888148       open     7308   
1752  https://id.biodiversity.org.au/node/apni/2920995       open     7308   
1761  https://id.biodiversity.org.au/node/apni/2897499       open     7308   
...                                                ...        ...      ...   
2283  https://id.biodiversity.org.au/node/apni/2895613       open     7308   
2286  https://id.biodiversity.org.au/node/apni/2897589       open     7308   
2288  https://id.biodiversity.org.au/node/apni/2902255       open     7308   
2291                                    ALA_DR652_1994       open     7308   
2298  https://id.biodiversity.org.au/node/apni/2904113       open     7308   

       username                                        description  
1749  peggydnew  Listed as Confidential - refer to https://www....  
1750  peggydnew  Listed as Confidential - refer to https://www....  
1751  peggydnew  Listed as Confidential - refer to https://www....  
1752  peggydnew  Listed as Confidential - refer to https://www....  
1761  peggydnew  Listed as Confidential - refer to https://www....  
...         ...                                                ...  
2283  peggydnew  Listed as Confidential - refer to https://www....  
2286  peggydnew  Listed as Confidential - refer to https://www....  
2288  peggydnew  Listed as Confidential - refer to https://www....  
2291  peggydnew  Listed as Confidential - refer to https://www....  
2298  peggydnew  Listed as Confidential - refer to https://www....  

[574 rows x 12 columns]

In [45]:
# investigation - which updates are mine (346), which are those not from me (80 not from me)
#updates[updates['user_id']=='708886'][['scientificName','new_status','status_inat','authority','new_authority','description','new_description','geoprivacy','new_geoprivacy']]
#updates[updates['user_id']!='708886'][['user_id','scientificName','new_status','status_inat','authority','new_authority','description','new_description','geoprivacy','new_geoprivacy']]
# those with different statuses
#updates[updates['new_status'].str.lower().str.strip() != updates['status_inat'].str.lower().str.strip()][['scientificName','new_status','status_inat','authority','new_authority','description','new_description','geoprivacy','new_geoprivacy']]
# users who've updated qld statuses who aren't me
#'https://www.inaturalist.org/users/220795','Steven Kurniawidjaja','neontetraploid','US'
#'https://www.inaturalist.org/users/3669610','Craig Robbins','craig-r','AU'
#'https://www.inaturalist.org/users/527710','James Kameron Mitchell','jameskm','US'
#'https://www.inaturalist.org/users/58320','lwnrngr','lwnrngr','NZ'
#'https://www.inaturalist.org/users/702203','Kitty Maurey','kitty12','CA'
#'https://www.inaturalist.org/users/717122','Miguel de Salas','mftasp','TAS'


## No status in iNaturalist via straight scientificName match
The Qld records that didn't match up to a status in iNaturalist

In [46]:
# to add: those that have no inaturalist status - 532!!
noinatstatus = mergedstatuses[mergedstatuses['status_id'].isnull()]
# try to match the taxon name to something in inaturalist
noinatstatus = noinatstatus.merge(inattaxa, how="left", left_on="scientificName",right_on="scientificName")
noinatstatus

wildnetTaxonID        scientificName                 status  \
0             33360   Abrodictyum brassii  Special least concern   
1             33358  Abrodictyum caudatum  Special least concern   
2             33359  Abrodictyum obscurum  Special least concern   
3             11892    Acacia acrionastes        Near Threatened   
4             14898      Acacia ammophila             Vulnerable   
...             ...                   ...                    ...   
1910          28217         Zieria vagans  Critically Endangered   
1911           3296      Zieria verrucosa             Vulnerable   
1912          41054    Zieria wilhelminae  Critically Endangered   
1913          14130    Zostera capricorni  Special least concern   
1914            876                   NaN             Endangered   

     new_geoprivacy                                               lsid  \
0              open   https://id.biodiversity.org.au/node/apni/7401847   
1              open   https://id.biodiversity.org.au/node/apni/7402200   
2              open   https://id.biodiversity.org.au/node/apni/7402565   
3              open   https://id.biodiversity.org.au/node/apni/2900521   
4              open   https://id.biodiversity.org.au/node/apni/2899480   
...             ...                                                ...   
1910           open   https://id.biodiversity.org.au/node/apni/2900138   
1911           open  https://id.biodiversity.org.au/taxon/apni/5136...   
1912           open                                     ALA_DR652_2004   
1913           open   https://id.biodiversity.org.au/node/apni/2905201   
1914           open                                      ALA_DR652_233   

     status_id taxon_id user_id description iucn  ...        phylum  \
0          NaN      NaN     NaN         NaN  NaN  ...           NaN   
1          NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
2          NaN      NaN     NaN         NaN  NaN  ...           NaN   
3          NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
4          NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
...        ...      ...     ...         ...  ...  ...           ...   
1910       NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
1911       NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
1912       NaN      NaN     NaN         NaN  NaN  ...           NaN   
1913       NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
1914       NaN      NaN     NaN         NaN  NaN  ...           NaN   

               class            order            family        genus  \
0                NaN              NaN               NaN          NaN   
1     Polypodiopsida  Hymenophyllales  Hymenophyllaceae  Abrodictyum   
2                NaN              NaN               NaN          NaN   
3      Magnoliopsida          Fabales          Fabaceae       Acacia   
4      Magnoliopsida          Fabales          Fabaceae       Acacia   
...              ...              ...               ...          ...   
1910   Magnoliopsida       Sapindales          Rutaceae       Zieria   
1911   Magnoliopsida       Sapindales          Rutaceae       Zieria   
1912             NaN              NaN               NaN          NaN   
1913      Liliopsida      Alismatales       Zosteraceae      Zostera   
1914             NaN              NaN               NaN          NaN   

     specificEpithet infraspecificEpithet              modified taxonRank  \
0                NaN                  NaN                   NaN       NaN   
1           caudatum                  NaN  2022-06-07T08:14:14Z   species   
2                NaN                  NaN                   NaN       NaN   
3        acrionastes                  NaN  2022-04-06T22:12:50Z   species   
4          ammophila                  NaN  2022-04-07T01:03:28Z   species   
...              ...                  ...                   ...       ...   
1910          vagans                  NaN 

In [47]:
noinatstatus[noinatstatus['id'].notna()] # there's no status but there is a matching inat taxon (id is the taxon id)
# note: "Dendrobium" matches to both genus and section

wildnetTaxonID        scientificName                 status  \
1             33358  Abrodictyum caudatum  Special least concern   
3             11892    Acacia acrionastes        Near Threatened   
4             14898      Acacia ammophila             Vulnerable   
5              3304        Acacia arbiana        Near Threatened   
6             31097      Acacia argentina             Vulnerable   
...             ...                   ...                    ...   
1908          27791        Zieria obovata             Vulnerable   
1909           9476       Zieria rimulosa             Vulnerable   
1910          28217         Zieria vagans  Critically Endangered   
1911           3296      Zieria verrucosa             Vulnerable   
1913          14130    Zostera capricorni  Special least concern   

     new_geoprivacy                                               lsid  \
1              open   https://id.biodiversity.org.au/node/apni/7402200   
3              open   https://id.biodiversity.org.au/node/apni/2900521   
4              open   https://id.biodiversity.org.au/node/apni/2899480   
5              open   https://id.biodiversity.org.au/node/apni/2906686   
6              open   https://id.biodiversity.org.au/node/apni/2888490   
...             ...                                                ...   
1908           open   https://id.biodiversity.org.au/node/apni/2917229   
1909           open   https://id.biodiversity.org.au/node/apni/2913391   
1910           open   https://id.biodiversity.org.au/node/apni/2900138   
1911           open  https://id.biodiversity.org.au/taxon/apni/5136...   
1913           open   https://id.biodiversity.org.au/node/apni/2905201   

     status_id taxon_id user_id description iucn  ...        phylum  \
1          NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
3          NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
4          NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
5          NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
6          NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
...        ...      ...     ...         ...  ...  ...           ...   
1908       NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
1909       NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
1910       NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
1911       NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   
1913       NaN      NaN     NaN         NaN  NaN  ...  Tracheophyta   

               class            order            family        genus  \
1     Polypodiopsida  Hymenophyllales  Hymenophyllaceae  Abrodictyum   
3      Magnoliopsida          Fabales          Fabaceae       Acacia   
4      Magnoliopsida          Fabales          Fabaceae       Acacia   
5      Magnoliopsida          Fabales          Fabaceae       Acacia   
6      Magnoliopsida          Fabales          Fabaceae       Acacia   
...              ...              ...               ...          ...   
1908   Magnoliopsida       Sapindales          Rutaceae       Zieria   
1909   Magnoliopsida       Sapindales          Rutaceae       Zieria   
1910   Magnoliopsida       Sapindales          Rutaceae       Zieria   
1911   Magnoliopsida       Sapindales          Rutaceae       Zieria   
1913      Liliopsida      Alismatales       Zosteraceae      Zostera   

     specificEpithet infraspecificEpithet              modified taxonRank  \
1           caudatum                  NaN  2022-06-07T08:14:14Z   species   
3        acrionastes                  NaN  2022-04-06T22:12:50Z   species   
4          ammophila                  NaN  2022-04-07T01:03:28Z   species   
5            arbiana                  NaN  2022-04-06T22:14:00Z   species   
6          argentina                  NaN  2022-04-03T23:48:35Z   species   
...              ...                  ...                   ...       ...   
1908         obovata                  NaN 

In [48]:
# there's no status but there is a matching inat taxon (id is the taxon id)
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions.sort_values(['scientificName'])
additions['action'] = 'ADD'
additions = additions[['action','scientificName','status_id','id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
additions.columns = additions.columns.str.replace("new_", "", regex=True)
additions = additions.rename(columns={'scientificName':'taxon_name',
                                      'id':'taxon_id',
                                  'status_id':'id'})
additions

action            taxon_name   id taxon_id                 status  \
1       ADD  Abrodictyum caudatum  NaN   451374  Special least concern   
3       ADD    Acacia acrionastes  NaN   898576        Near Threatened   
4       ADD      Acacia ammophila  NaN   898600             Vulnerable   
5       ADD        Acacia arbiana  NaN   898626        Near Threatened   
6       ADD      Acacia argentina  NaN   898630             Vulnerable   
...     ...                   ...  ...      ...                    ...   
1908    ADD        Zieria obovata  NaN   956303             Vulnerable   
1909    ADD       Zieria rimulosa  NaN  1247621             Vulnerable   
1910    ADD         Zieria vagans  NaN  1247619  Critically Endangered   
1911    ADD      Zieria verrucosa  NaN  1247620             Vulnerable   
1913    ADD    Zostera capricorni  NaN   407068  Special least concern   

     iucn_equivalent                                  authority  \
1         Vulnerable  Qld Department of Environment and Science   
3         Vulnerable  Qld Department of Environment and Science   
4         Vulnerable  Qld Department of Environment and Science   
5         Vulnerable  Qld Department of Environment and Science   
6         Vulnerable  Qld Department of Environment and Science   
...              ...                                        ...   
1908      Vulnerable  Qld Department of Environment and Science   
1909      Vulnerable  Qld Department of Environment and Science   
1910      Vulnerable  Qld Department of Environment and Science   
1911      Vulnerable  Qld Department of Environment and Science   
1913      Vulnerable  Qld Department of Environment and Science   

                                                    url geoprivacy place_id  \
1      https://id.biodiversity.org.au/node/apni/7402200       open     7308   
3      https://id.biodiversity.org.au/node/apni/2900521       open     7308   
4      https://id.biodiversity.org.au/node/apni/2899480       open     7308   
5      https://id.biodiversity.org.au/node/apni/2906686       open     7308   
6      https://id.biodiversity.org.au/node/apni/2888490       open     7308   
...                                                 ...        ...      ...   
1908   https://id.biodiversity.org.au/node/apni/2917229       open     7308   
1909   https://id.biodiversity.org.au/node/apni/2913391       open     7308   
1910   https://id.biodiversity.org.au/node/apni/2900138       open     7308   
1911  https://id.biodiversity.org.au/taxon/apni/5136...       open     7308   
1913   https://id.biodiversity.org.au/node/apni/2905201       open     7308   

       username                                        description  
1     peggydnew  Listed as Confidential - refer to https://www....  
3     peggydnew  Listed as Confidential - refer to https://www....  
4     peggydnew  Listed as Confidential - refer to https://www....  
5     peggydnew  Listed as Confidential - refer to https://www....  
6     peggydnew  Listed as Confidential - refer to https://www....  
...         ...                                                ...  
1908  peggydnew  Listed as Confidential - refer to https://www....  
1909  peggydnew  Listed as Confidential - refer to https://www....  
1910  peggydnew  Listed as Confidential - refer to https://www....  
1911  peggydnew  Listed as Confidential - refer to https://www....  
1913  peggydnew  Listed as Confidential - refer to https://www....  

[1367 rows x 12 columns]

In [49]:
# write these to the file
pd.concat([updates, additions]).to_csv(sourcedir + "qld.csv", index=False)

In [50]:
# what didnt match to a taxon?
unknownToInat = noinatstatus[noinatstatus['id'].isna()]
unknownToInat

wildnetTaxonID                 scientificName                 status  \
0             33360            Abrodictyum brassii  Special least concern   
2             33359           Abrodictyum obscurum  Special least concern   
9             41056                Acacia castorum             Vulnerable   
14            40993                Acacia forsteri  Critically Endangered   
19            14862                   Acacia lauta             Vulnerable   
...             ...                            ...                    ...   
1890          41972                Xylosma craynii             Vulnerable   
1893          41347  Zealandia pustulata pustulata  Special least concern   
1895          40091              Zeuxine attenuata          Least concern   
1912          41054             Zieria wilhelminae  Critically Endangered   
1914            876                            NaN             Endangered   

     new_geoprivacy                                               lsid  \
0              open   https://id.biodiversity.org.au/node/apni/7401847   
2              open   https://id.biodiversity.org.au/node/apni/7402565   
9              open                                      ALA_DR652_969   
14             open                                      ALA_DR652_974   
19             open   https://id.biodiversity.org.au/node/apni/2886854   
...             ...                                                ...   
1890           open   https://id.biodiversity.org.au/node/apni/2903337   
1893           open                                     ALA_DR652_1822   
1895       obscured  https://id.biodiversity.org.au/taxon/apni/5140...   
1912           open                                     ALA_DR652_2004   
1914           open                                      ALA_DR652_233   

     status_id taxon_id user_id description iucn  ... phylum class order  \
0          NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
2          NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
9          NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
14         NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
19         NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
...        ...      ...     ...         ...  ...  ...    ...   ...   ...   
1890       NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
1893       NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
1895       NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
1912       NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
1914       NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   

     family genus specificEpithet infraspecificEpithet modified taxonRank  \
0       NaN   NaN             NaN                  NaN      NaN       NaN   
2       NaN   NaN             NaN                  NaN      NaN       NaN   
9       NaN   NaN             NaN                  NaN      NaN       NaN   
14      NaN   NaN             NaN                  NaN      NaN       NaN   
19      NaN   NaN             NaN                  NaN      NaN       NaN   
...     ...   ...             ...                  ...      ...       ...   
1890    NaN   NaN             NaN                  NaN      NaN       NaN   
1893    NaN   NaN             NaN                  NaN      NaN       NaN   
1895    NaN   NaN             NaN                  NaN      NaN       NaN   
1912    NaN   NaN             NaN                  NaN      NaN       NaN   
1914    NaN   NaN             NaN                  NaN      NaN       NaN   

     references  
0           NaN  
2           NaN  
9           NaN  
14          NaN  
19          NaN  
...         ...  
1890        NaN  
1893        NaN  
1895        NaN  
1912        NaN  
1914        NaN  

[548 rows x 37 columns]

In [51]:
noinatstatus.groupby('status').size()

status
Critically Endangered    101
Endangered               217
Extinct                    5
Extinct in the wild       23
Least concern             81
Near Threatened          215
Special least concern    730
Vulnerable               457
dtype: int64

In [52]:
# what didnt match to a taxon?
noinatstatus[noinatstatus['id'].isna()]

wildnetTaxonID                 scientificName                 status  \
0             33360            Abrodictyum brassii  Special least concern   
2             33359           Abrodictyum obscurum  Special least concern   
9             41056                Acacia castorum             Vulnerable   
14            40993                Acacia forsteri  Critically Endangered   
19            14862                   Acacia lauta             Vulnerable   
...             ...                            ...                    ...   
1890          41972                Xylosma craynii             Vulnerable   
1893          41347  Zealandia pustulata pustulata  Special least concern   
1895          40091              Zeuxine attenuata          Least concern   
1912          41054             Zieria wilhelminae  Critically Endangered   
1914            876                            NaN             Endangered   

     new_geoprivacy                                               lsid  \
0              open   https://id.biodiversity.org.au/node/apni/7401847   
2              open   https://id.biodiversity.org.au/node/apni/7402565   
9              open                                      ALA_DR652_969   
14             open                                      ALA_DR652_974   
19             open   https://id.biodiversity.org.au/node/apni/2886854   
...             ...                                                ...   
1890           open   https://id.biodiversity.org.au/node/apni/2903337   
1893           open                                     ALA_DR652_1822   
1895       obscured  https://id.biodiversity.org.au/taxon/apni/5140...   
1912           open                                     ALA_DR652_2004   
1914           open                                      ALA_DR652_233   

     status_id taxon_id user_id description iucn  ... phylum class order  \
0          NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
2          NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
9          NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
14         NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
19         NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
...        ...      ...     ...         ...  ...  ...    ...   ...   ...   
1890       NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
1893       NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
1895       NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
1912       NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   
1914       NaN      NaN     NaN         NaN  NaN  ...    NaN   NaN   NaN   

     family genus specificEpithet infraspecificEpithet modified taxonRank  \
0       NaN   NaN             NaN                  NaN      NaN       NaN   
2       NaN   NaN             NaN                  NaN      NaN       NaN   
9       NaN   NaN             NaN                  NaN      NaN       NaN   
14      NaN   NaN             NaN                  NaN      NaN       NaN   
19      NaN   NaN             NaN                  NaN      NaN       NaN   
...     ...   ...             ...                  ...      ...       ...   
1890    NaN   NaN             NaN                  NaN      NaN       NaN   
1893    NaN   NaN             NaN                  NaN      NaN       NaN   
1895    NaN   NaN             NaN                  NaN      NaN       NaN   
1912    NaN   NaN             NaN                  NaN      NaN       NaN   
1914    NaN   NaN             NaN                  NaN      NaN       NaN   

     references  
0           NaN  
2           NaN  
9           NaN  
14          NaN  
19          NaN  
...         ...  
1890        NaN  
1893        NaN  
1895        NaN  
1912        NaN  
1914        NaN  

[548 rows x 37 columns]

### are there any that need to be removed?
qld list count: 2517
qld inat statuses count: 653

updates to inat status: 570
additional inat status: 1355
qld statuses we can't find a taxon match for in iNaturalist: 606
total: 2531 (explainable via the various genus/section entries that we matched to in the taxonomy)

inat statuses left over: 653-570=83 that may need checking against the above

In [53]:
# inat statuses that aren't in added or updated
notaddedupdated = inatstatuses[~inatstatuses['taxon_id'].isin(updates['taxon_id'])]
notaddedupdated

status_id taxon_id  user_id place_id source_id  \
102     159922  1019990   702203     7308             
437     234929   104980  3669610     7308             
18      167722  1073668  3669610     7308             
308     264555  1074506  3669610     7308             
2614    161805  1091300   702203     7308             
...        ...      ...      ...      ...       ...   
897     154451   937266   708886     7308     16653   
837     154918   954413   708886     7308     16653   
2815    264331   961868  3669610     7308             
1053    155106   963748   708886     7308     16653   
23      167723   993605  3669610     7308             

                                             authority      status  \
102                                         Queensland          VU   
437            Queensland Nature Conservation Act 1992          VU   
18                               Queensland Government          LC   
308   Queensland Department of Environment and Science          LC   
2614                      Nature Conservation Act 1992          EN   
...                                                ...         ...   
897                                           QLD DEHP  endangered   
837                                           QLD DEHP  endangered   
2815           Queensland Nature Conservation Act 1992          LC   
1053                                          QLD DEHP  endangered   
23                                            QLD DEHP          NT   

                                                    url  \
102   https://www.legislation.qld.gov.au/view/html/i...   
437   https://apps.des.qld.gov.au/species-search/det...   
18    https://www.data.qld.gov.au/dataset/conservati...   
308   https://apps.des.qld.gov.au/species-search/det...   
2614  http://155.187.2.69/cgi-bin/sprat/public/publi...   
...                                                 ...   
897   https://data.qld.gov.au/dataset/conservation-s...   
837   https://data.qld.gov.au/dataset/conservation-s...   
2815  https://apps.des.qld.gov.au/species-search/det...   
1053  https://data.qld.gov.au/dataset/conservation-s...   
23    https://www.data.qld.gov.au/dataset/conservati...   

                                            description geoprivacy  ...  \
102                                                                 ...   
437                                                                 ...   
18    Least concern, locations still considered conf...             ...   
308                                                           open  ...   
2614                                                                ...   
...                                                 ...        ...  ...   
897                                                       obscured  ...   
837                                                       obscured  ...   
2815                                                          open  ...   
1053                                                      obscured  ...   
23                                                        obscured  ...   

               genus specificEpithet infraspecificEpithet  \
102           Acacia          baueri                        
437   Maccullochella          peelii                        
18           Liparis   coelogynoides                        
308       Eucalyptus     chloroclada                        
2614                                                        
...              ...             ...                  ...   
897                                                         
837                                                         
2815      Eucalyptus      thozetiana                        
1053        Nervilia         simplex                        
23         Acianthus   amplexicaulis                        

                  modified             scientificName taxonRank  \
102   2022-04-06T22:03:46Z              Acacia baueri   species   
437   2022-01-28T04:34:59Z

In [54]:
# Stats
numsensitive = len(sensitivelist.index)
numconservation = len(conservationlist.index)
numupdates  = len(updates.index)
numadditions  = len(additions.index)
numnoinatstatus = len(noinatstatus.index)
numunknownToInat = len(unknownToInat.index)
numnotaddedupdated = len(notaddedupdated.index)
numnoncomply = len(noncomply.index)
numcomply = len(statelist.index)
numdupinfo = len(dupinformation.index)
d = {'Sensitive': [numsensitive],
    'Conservation': [numconservation],
    'Statelist merge': [numfullstatelist],
    'Species iNat Comply' : [numcomply],
    'Species iNat non-Comply': [numnoncomply],
    'Duplicate Information': [numdupinfo],
    'Updates': [numupdates],
    'Additions': [numadditions],
    'Not added updated': [numnotaddedupdated],
    'No Inat Status': [numnoinatstatus],
    'Unknown to Inat': [numunknownToInat]}

statsdf = pd.DataFrame(data=d)
statsdf

Sensitive  Conservation  Statelist merge  Species iNat Comply  \
0        955          2244             2517                 2473   

   Species iNat non-Comply  Duplicate Information  Updates  Additions  \
0                       44                      0      574       1367   

   Not added updated  No Inat Status  Unknown to Inat  
0                 81            1915              548